[jit] make wildcards alias only each other

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#20670 [jit] make wildcards alias only each other**
* #20556 [jit] Mark values entering containers as wildcards

Context found at the bottom of this stack. This PR makes wildcards alias
only each other (instead of any value in the graph). Special "wildcard"
memoryDAG elements are created to model this aliasing relationship.
Graph inputs also now alias the same wildcard set.

Since the rules for wildcards aren't special anymore, we can eliminate a
ton of special-casing code for wildcards!

Differential Revision: [D15447567](https://our.internmc.facebook.com/intern/diff/D15447567)

suo committed:
[jit] make wildcards alias only each other

Context found at the bottom of this stack. This PR makes wildcards alias
only each other (instead of any value in the graph). Special "wildcard"
memoryDAG elements are created to model this aliasing relationship.
Graph inputs also now alias the same wildcard set.

Since the rules for wildcards aren't special anymore, we can eliminate a
ton of special-casing code for wildcards!

eellison reviewed (commented):
Looks really good!

I think you need to remove the following lines in alias_analysis.h
```
if std::any_of(b.cbegin(), b.cend(), [this](const Value* v) {
  return isWildcard(v);
})) {
      return true;
    }
```

suo commented on torch/csrc/jit/passes/alias_analysis.cpp:
```diff
@@ -9,15 +9,40 @@
+      return TypeKind::DictType;
+    case TypeKind::VarType:
+      return TypeKind::VarType;
+    case TypeKind::FutureType:
```
I realize now that futures are not mutable so they don't need alias annotations. So I'll remove them entirely

suo commented:
The lines you pointed out have been removed?

suo commented on torch/csrc/jit/passes/alias_analysis.cpp:
```diff
@@ -279,75 +248,10 @@ void AliasDb::dump() const {
-// TODO: need to create a dummy "graph input alias" value in MemoryDAG for all
-// inputs of the same type to point to. Currently they all point to the first
-// element, which is technically wrong.
-void AliasDb::makeAllAlias(const std::vector<Value*>& values) {
```
I think that would kind of run in the other direction from where we want to go. We still want to keep "outside this function" as the wildcard set, but we should work toward adding refinements for "things that are inside the function" like container creation and extraction.

eellison commented:
> The lines you pointed out have been removed?

Oop that comment myst have been from first commit

zdevito commented on torch/csrc/jit/passes/alias_analysis.cpp:
```diff
@@ -1400,13 +1249,30 @@ void AliasDb::rebuildWriteCache() const {
 }
 
-ValueSet AliasDb::getMemoryLocations(const Value* v) const {
+// Given a future value `v`, figure out what the "original" futures that
```
Discussed a bit in person. It doesn't seem like this can always be accurate, since it possible to do, e.g.
```
a = [fut]
if <halting problem>:
  a.append(fut2)
b = a.pop(rand(2)).wait()
```
So futures cannot always be traced to a fork operation. Handling them correctly seems similar to better analysis of what may/may not be in a list. So maybe we should just implement both at the same time in another PR. For now, we can treat it like list extraction.

zdevito commented on torch/csrc/jit/passes/alias_analysis.h:
```diff
@@ -241,17 +210,18 @@ class AliasDb {
+  std::map<TypeKind, Element*> wildcardIndex_;
+  Element* getWildcard(const TypePtr& type) const;
+  Element* getOrCreateWildcard(const TypePtr& type);
+  bool isWildcard(const Value* v) const;
```
isWildcard might be better as mayAliasWildcard

zdevito reviewed (approved):
Yessss. I like how this simplifies this and sets us up for doing better alias tracking through Tuple, Lists, futures.

suo committed:
Update on "[jit] make wildcards alias only each other"

[jit] make wildcards alias only each other

Context found at the bottom of this stack. This PR makes wildcards alias
only each other (instead of any value in the graph). Special "wildcard"
memoryDAG elements are created to model this aliasing relationship.
Graph inputs also now alias the same wildcard set.

Since the rules for wildcards aren't special anymore, we can eliminate a
ton of special-casing code for wildcards!

gh-metadata: pytorch pytorch 20670 gh/suo/40/head

suo committed:
Update on "[jit] make wildcards alias only each other"

[jit] make wildcards alias only each other

Context found at the bottom of this stack. This PR makes wildcards alias
only each other (instead of any value in the graph). Special "wildcard"
memoryDAG elements are created to model this aliasing relationship.
Graph inputs also now alias the same wildcard set.

Since the rules for wildcards aren't special anymore, we can eliminate a
ton of special-casing code for wildcards!

gh-metadata: pytorch pytorch 20670 gh/suo/40/head
